web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState, useEffect, useRef } from 'react';
+import useAttempt, { Attempt } from 'shared/hooks/useAttemptNext';
+
+import {
+  ResourcesResponse,
+  ResourceFilter,
+  resourceFilterToHookDeps,
+} from 'teleport/services/agents';
+import { UrlResourcesParams } from 'teleport/config';
+
+/**
+ * Supports fetching more data from the server when more data is available. Pass
+ * a `fetchFunc` that retrieves a single batch of data. After the initial
+ * request, the server is expected to return a `startKey` field that denotes the
+ * next `startKey` to use for the next request.
+ *
+ * This hook is intended to be used in tandem with the `useInfiniteScroll` hook.
+ * See its documentation for an example.
+ */
+export function useKeyBasedPagination<T>({
+  fetchFunc,
+  clusterId,
+  filter,
+  initialFetchSize = 30,
+  fetchMoreSize = 20,
+}: Props<T>): State<T> {
+  const abortController = useRef<AbortController | null>(null);
+  const { attempt, setAttempt } = useAttempt();
+  const [finished, setFinished] = useState(false);
+  const [resources, setResources] = useState<T[]>([]);
+  const [startKey, setStartKey] = useState(null);
+
+  useEffect(() => {
+    abortController.current?.abort();
+    abortController.current = null;
+
+    setAttempt({ status: '', statusText: '' });
+    setFinished(false);
+    setResources([]);
+    setStartKey(null);
+  }, [clusterId, ...resourceFilterToHookDeps(filter)]);
```

Comment:
Yes and no. Yes, it may theoretically be sufficient; no, because since changing filters resets the pagination. In this case, I'd rather compare a couple of fields than risk that a place that decides to parse something from an URL creates two independent instances of filters with the same semantics, but different referential identity.

Reply:
I think I'm gathering that the `filter` is parsed from the URL, and that changing the `filter` resets the pagination, hence the need to reset the state of this hook whenever the `filter` changes (same goes for `clusterId`).

The reason you compare fields directly is that you want to account for a case where the upstream URL parsing code is triggered to create different objects for `filter` but with the same fields -- in this case we won't trigger these effects unnecessarily, as compared to just passing in `filter` directly.

That makes mechanical/semantic sense, but it nevertheless seems an odd piece to optimize. AFAIK there's no reason to think that scenario should occur frequently, and if it were to occur the component should be robust to it and the performance impact should be negligible. So my opinion is still that you should just get rid of it, but if you disagree (or I'm still in some way confused), a comment on `resourceFilterToHookDeps` explaining precisely what's being optimized for will be useful for understanding this months down the line.

Reply:
I get your point here. I was trying to prevent sneaky bugs, but this one would actually be pretty obvious (resetting the view out of nowhere). I'm gonna remove this.